src/plugins/ipt/ipt.cpp

```diff
@@ -0,0 +1,469 @@
+ * on https://github.com/tklengyel/drakvuf, or by other methods.           *
+ * By sending these changes, it is understood (unless you specify          *
+ * otherwise) that you are offering unlimited, non-exclusive right to      *
+ * reuse, modify, and relicense the code.  DRAKVUF will always be          *
+ * available Open Source, but this is important because the inability to   *
+ * relicense code has caused devastating problems for other Free Software  *
+ * projects (such as KDE and NASM).                                        *
+ * To specify special license conditions of your contributions, just say   *
+ * so when you send them.                                                  *
+ *                                                                         *
+ * This program is distributed in the hope that it will be useful, but     *
+ * WITHOUT ANY WARRANTY; without even the implied warranty of              *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the DRAKVUF   *
+ * license file for more details (it's in a COPYING file included with     *
+ * DRAKVUF, and also available from                                        *
+ * https://github.com/tklengyel/drakvuf/COPYING)                           *
+ *                                                                         *
+ ***************************************************************************/
+
+#include <config.h>
+#include <glib.h>
+#include <inttypes.h>
+#include <sys/stat.h>
+#include <libvmi/libvmi.h>
+#include <libvmi/peparse.h>
+#include <assert.h>
+#include <libdrakvuf/json-util.h>
+#include <set>
+
+#include "ipt.h"
+#include "plugins/output_format.h"
+#include "private.h"
+
+
+/*
+ * Dump current IPT logs to disk and annotate them with custom PTWRITE packets
+ */
+static inline
+int annotate_ipt(drakvuf_t drakvuf, ipt* plugin, unsigned int vcpu_id, uint32_t* payloads, size_t num_payloads)
+{
+    uint64_t offset;
+    uint64_t last_offset;
+    bool ret = drakvuf_get_ipt_offset(drakvuf, vcpu_id, &offset, &last_offset);
+    int fail = 0;
+
+    uint8_t* buf = plugin->vcpu[vcpu_id].buf;
+    uint64_t size = plugin->vcpu[vcpu_id].size;
+    FILE* fd = plugin->vcpu[vcpu_id].fd;
+
+    if (!ret)
+    {
+        PRINT_DEBUG("annotate_ipt() failed to get ipt offset for vcpu %d\n", vcpu_id);
+        return 0;
+    }
+
+    PRINT_DEBUG("annotate_ipt() vCPU: %d IPT_CUR: %llx IPT_LAST: %llx\n", vcpu_id, (unsigned long long)offset, (unsigned long long)last_offset);
+
+    if (offset > last_offset)
+    {
+        fwrite(buf + last_offset, offset - last_offset, 1, fd);
+    }
+    else if (offset < last_offset)
+    {
+        fwrite(buf + last_offset, size - last_offset, 1, fd);
+        fwrite(buf, offset, 1, fd);
+    }
+    else
+    {
+        PRINT_DEBUG("annotate_ipt() called but no new IPT data is present\n");
+        fail = 1;
+    }
+
+    uint8_t ptwrite_packet[10] = {0x02, 0x32,};
+    uint32_t x1 = PTW_ERROR_EMPTY;
+    uint32_t x2 = 0;
```

Comment:
these are reassigned every for loop AND in the if case, so what's the point in early declaration here? Also `x1` and `x2` are pretty enigmatic names.